Cython/Compiler/Optimize.py

```diff
@@ -333,6 +339,78 @@ def _transform_reversed_iteration(self, node, reversed_function):
             PyrexTypes.CFuncTypeArg("s", Builtin.bytes_type, None)
             ])
 
+    def _transform_indexable_iteration(self, node, slice_node, reversed=False):
+        """In principle can handle any iterable that Cython has a len() for and knows how to index"""
+        unpack_temp_node = UtilNodes.LetRefNode(
+            slice_node.as_none_safe_node("'NoneType' is not iterable"),
+            may_hold_none=False
+            )
+
+        start_node = ExprNodes.IntNode(
+            node.pos, value='0', constant_result=0, type=PyrexTypes.c_py_ssize_t_type)
+        length_temp = UtilNodes.LetRefNode(ExprNodes.SimpleCallNode(node.pos,
+                                    function = ExprNodes.NameNode(node.pos, name="len"),
+                                    args = [unpack_temp_node]
+                                    ), type=PyrexTypes.c_py_ssize_t_type)
+        end_node = length_temp
+        if reversed:
+            relation1, relation2 = '>', '>='
+            start_node, end_node = end_node, start_node
+        else:
+            relation1, relation2 = '<=', '<'
+
+        counter_ref = UtilNodes.LetRefNode(pos=node.pos, type=PyrexTypes.c_py_ssize_t_type)
+
+        target_value = ExprNodes.IndexNode(slice_node.pos, base=unpack_temp_node,
+                                           index=counter_ref)
+
+        target_assign = Nodes.SingleAssignmentNode(
+            pos = node.target.pos,
+            lhs = node.target,
+            rhs = target_value)
+
+        # analyse with boundscheck and wraparound
+        # off (because we're confident we know the size)
+        env = self.current_env()
+        new_directives = dict(env.directives, boundscheck=False, wraparound=False)
+        new_directives.pop('test_assert_path_exists', None)
+        new_directives.pop('test_fail_if_path_exists', None)
+        target_assign = Nodes.CompilerDirectivesNode(target_assign.pos,
+                                                        directives = new_directives,
+                                                        body = target_assign)
```

Comment:
There are a couple of other places in this PR where more consistent code formatting would help the readability.
Sorry, but when I see a red door, I want it painted black. ;)
```suggestion
        target_assign = Nodes.CompilerDirectivesNode(
            target_assign.pos,
            directives=new_directives,
            body=target_assign,
        )
```